pkg/recovery/apiserver.go

```diff
@@ -0,0 +1,438 @@
+	if err != nil {
+		return fmt.Errorf("failed to create client certificate: %v", err)
+	}
+
+	clientCertBytes, clientKeyBytes, err := clientCert.GetPEMBytes()
+
+	s.restConfig = &rest.Config{
+		Host: "https://localhost:7443",
+		TLSClientConfig: rest.TLSClientConfig{
+			CAFile:   servingCaCertPath,
+			CertData: clientCertBytes,
+			KeyData:  clientKeyBytes,
+		},
+	}
+
+	kubeconfig, err := s.KubeConfig()
+	if err != nil {
+		return fmt.Errorf("failed to create kubeconfig: %v", err)
+	}
+
+	kubeconfigBytes, err := yaml.Marshal(kubeconfig)
+	if err != nil {
+		return fmt.Errorf("failed to marshal kubeconfig: %v", err)
+	}
+
+	kubeconfigPath := path.Join(recoveryDirPath, AdminKubeconfigFileName)
+	err = ioutil.WriteFile(kubeconfigPath, kubeconfigBytes, 600)
+	if err != nil {
+		return fmt.Errorf("failed to write kubeconfig %q: %v", kubeconfigPath, err)
+	}
+
+	return nil
+}
+
+func (s *Apiserver) WaitForHealthz(ctx context.Context) error {
+	// We don't have to worry about connecting to an old, terminating apiserver
+	// as our client certs are unique to this instance
+
+	kubeClientset, err := s.GetKubeClientset()
+	if err != nil {
+		return err
+	}
+
+	return wait.PollUntil(500*time.Millisecond, func() (bool, error) {
+		req := kubeClientset.RESTClient().Get().AbsPath("/healthz")
+
+		klog.V(1).Infof("Waiting for recovery apiserver to come up at %q", req.URL())
+
+		_, err = req.DoRaw()
+		if err != nil {
+			klog.V(5).Infof("apiserver returned error: %v", err)
+			return false, nil
+		}
+
+		return true, nil
+	}, ctx.Done())
+}
+
+func (s *Apiserver) Destroy() error {
+	recoveryPodManifestPath := path.Join(s.PodManifestDir, RecoveryPodFileName)
+	err := os.Remove(recoveryPodManifestPath)
+	if err != nil {
+		return fmt.Errorf("failed to remove recovery pod manifest %q: %v", recoveryPodManifestPath, err)
+	}
+
+	// TODO: consider removing the recoveryDir
+	// (The dir is tmp by default anyways and putting a wrong path here could be fatal as it would have to recursively delete.)
+
+	return nil
+}
+
+func copyFile(src, dest string) error {
+	srcFile, err := os.Open(src)
+	if err != nil {
+		return fmt.Errorf("failed to open file %q: %v", src, err)
+	}
+	defer srcFile.Close()
+
+	srcFileInfo, err := srcFile.Stat()
+	if err != nil {
+		return fmt.Errorf("failed to stat file %q: %v", src, err)
+	}
+
+	if srcFileInfo.IsDir() {
+		return fmt.Errorf("can't copy file %q because it is a directory", src)
+	}
+
+	destFile, err := os.OpenFile(dest, os.O_WRONLY|os.O_CREATE|os.O_EXCL, srcFileInfo.Mode().Perm())
+	if err != nil {
+		return fmt.Errorf("failed to open file %q: %v", dest, err)
+	}
+	defer destFile.Close()
+
+	_, err = io.Copy(destFile, srcFile)
+	if err != nil {
+		return fmt.Errorf("failed to copy file %q into %q: %v", src, dest, err)
+	}
+
+	return nil
+}
+
+func getVolumeHostPathPath(volumeName string, volumes []corev1.Volume) (*string, error) {
```

Comment:
why a pointer? Unusual signature. Just return `""`.

Reply:
It's even dangerous as you can mutate the pod via this pointer. You probably (hopefully) don't want that.

Reply:
I was definitely doing just that at some iteration to change the hostpath while I was reusing the manifest, maybe it is not used now we have forked the manifest file